js/ai/src/prompt.ts

```diff
@@ -137,79 +174,381 @@ export interface ExecutablePrompt<
+export function definePrompt<
+  I extends z.ZodTypeAny = z.ZodTypeAny,
+  O extends z.ZodTypeAny = z.ZodTypeAny,
+  CustomOptions extends z.ZodTypeAny = z.ZodTypeAny,
+>(
   registry: Registry,
-  config: PromptConfig<I>,
-  fn: PromptFn<I>
-): PromptAction<I> {
-  const a = defineAction(
+  options: PromptConfig<I, O, CustomOptions>
+): ExecutablePrompt<z.infer<I>, O, CustomOptions> {
+  const promptCache = {} as PromptCache;
+
+  const renderOptionsFn = async (
+    input: z.infer<I>,
+    renderOptions: PromptGenerateOptions<O, CustomOptions> | undefined
+  ): Promise<GenerateOptions> => {
+    const messages: MessageData[] = [];
+    renderOptions = { ...renderOptions }; // make a copy, we will be trimming
+    const session = getCurrentSession(registry);
+
+    // order of these matters:
+    await renderSystemPrompt(
+      registry,
+      session,
+      input,
+      messages,
+      options,
+      promptCache
+    );
+    await renderMessages(
+      registry,
+      session,
+      input,
+      messages,
+      options,
+      renderOptions,
+      promptCache
+    );
+    await renderUserPrompt(
+      registry,
+      session,
+      input,
+      messages,
+      options,
+      promptCache
+    );
+
+    let docs: DocumentData[] | undefined;
+    if (typeof options.docs === 'function') {
+      docs = await options.docs(input, {
+        state: session?.state,
+      });
+    } else {
+      docs = options.docs;
+    }
+
+    return stripUndefinedProps({
+      model: options.model,
+      maxTurns: options.maxTurns,
+      messages,
+      docs,
+      tools: options.tools,
+      returnToolRequests: options.returnToolRequests,
+      toolChoice: options.toolChoice,
+      output: options.output,
+      use: options.use,
+      ...stripUndefinedProps(renderOptions),
+      config: {
+        ...options?.config,
+        ...renderOptions?.config,
+      },
+    });
+  };
+  const rendererAction = defineAction(
     registry,
     {
-      name: config.name,
-      inputJsonSchema: config.inputJsonSchema,
-      inputSchema: config.inputSchema,
-      description: config.description,
+      name: `${options.name}${options.variant ? `.${options.variant}` : ''}`,
+      inputJsonSchema: options.input?.jsonSchema,
+      inputSchema: options.input?.schema,
+      description: options.description,
       actionType: 'prompt',
-      metadata: { ...(config.metadata || { prompt: {} }), type: 'prompt' },
+      metadata: {
+        prompt: {
+          config: options.config,
+          input: {
+            schema: options.input ? toJsonSchema(options.input) : undefined,
+          },
+          name: options.name,
+          model: modelName(options.model),
+        },
+        ...options.metadata,
+        type: 'prompt',
+      },
     },
-    fn
+    async (input: z.infer<I>): Promise<GenerateRequest<z.ZodTypeAny>> => {
+      return toGenerateRequest(
+        registry,
+        await renderOptionsFn(input, undefined)
```

Comment:
Why an explicit `undefined` second argument? Seems like a smell.

Reply:
I got bitten by making it optional... when rendering from an action you actually can't override options, but in other places options are overridable. I made the arg explicit to make sure I don't forget to pass it somewhere where it is actually overridable.